File: 2. Process Scheduling Algorithm/FCFS (First Come First Serve)/FCFS.py
# Taking User Input
n = int(input("Enter number of jobs: "))
print("Enter process arrival time and burst time:")
jobs = []
for i in range(n):
    pt, att, btt = input().split()
    jobs.append((pt, int(att), int(btt)))

# Sorting
jobs.sort(key=lambda x: x[1])

# Extract sorted data into separate lists
p, at, bt, st, ct, tat, wt, rt = [], [], [], [], [], [], [], []
for job in jobs:
    p.append(job[0])
    at.append(job[1])
    bt.append(job[2])

# Calculate start time and completion time
t = 0
for i in range(n):
    if t < at[i]:
        t = at[i]
    st.append(t)
    t += bt[i]
    ct.append(t)

# Calculate tat, at, bt
for i in range(n):
    tat.append(ct[i]-at[i])
    wt.append(tat[i]-bt[i])
    rt.append(st[i]-at[i])

# Calculate totals
t_tat, t_wt, t_rt = sum(tat), sum(wt), sum(rt)

# Print Table
print("Process\tArrival Time\tBurst Time\tStart Time\tCompletion Time\tT.A.T\tW.T\tR.T")
for i in range(n):
    print(f"{p[i]}\t\t{at[i]}\t\t{bt[i]}\t\t{st[i]}\t\t{ct[i]}\t{tat[i]}\t{wt[i]}\t{rt[i]}")

print("Average turn around time:", t_tat/n)
print("Average waiting time:", t_wt/n)
print("Average response time:", t_rt/n)

File: 2. Process Scheduling Algorithm/FCFS (First Come First Serve)/FCFS.c
#include <stdio.h>
#include <stdlib.h>

struct Job {
    char p[5];
    int at, bt, st, ct, tat, wt, rt;
};

int compare(const void* a, const void* b) {
    return (*(struct Job*)a).at - (*(struct Job*)b).at;
}

int main() {
    // Taking User Input
    int n;
    printf("Enter number of jobs: ");
    scanf("%d", &n);
    struct Job arr[n];
    printf("Enter process arrival time and burst time\n");
    for (int i = 0; i < n; i++) scanf("%s %d %d", arr[i].p, &arr[i].at, &arr[i].bt);

    // Sorting
    qsort(arr, n, sizeof(struct Job), compare);

    // Calculate start time and completion time
    int t = 0;
    for (int i = 0; i < n; i++) {
        if (t < arr[i].at) t = arr[i].at;
        arr[i].st = t;
        t += arr[i].bt;
        arr[i].ct = t;
    }

    // Calculate tat, at, bt
    for (int i = 0; i < n; i++) {
        arr[i].tat = arr[i].ct - arr[i].at;
        arr[i].wt = arr[i].tat - arr[i].bt;
        arr[i].rt = arr[i].st - arr[i].at;
    }

    // Calculate totals
    float tat = 0, wt = 0, rt = 0;
    for (int i = 0; i < n; i++) {
        tat += arr[i].tat;
        wt += arr[i].wt;
        rt += arr[i].rt;
    }

    // Print Table
    printf("Process\tArrival Time\tBurst Time\tStart Time\tCompletion Time\tT.A.T\tW.T\tR.T\n");
    for (int i = 0; i < n; i++) {
        printf("%s\t%d\t\t%d\t\t%d\t\t%d\t\t%d\t%d\t%d\n", arr[i].p, arr[i].at, arr[i].bt, arr[i].st, arr[i].ct, arr[i].tat, arr[i].wt, arr[i].rt);
    }

    // Print Averages
    printf("Average turn around time: %.2f\n", tat / n);
    printf("Average waiting time: %.2f\n", wt / n);
    printf("Average response time: %.2f\n", rt / n);
    return 0;
}
